File: c/problem-2.c
#include <stdio.h>

#define MAX 4000000

int main() {
  // Project Euler states to start with 1 and 2.
  int sum = 2, pre1 = 1, pre2 = 2;
  int next;

  while (pre2 <= MAX) {
    next = pre1 + pre2;

    if (next % 2 == 0) {
      sum = sum + next;
    }

    pre1 = pre2;
    pre2 = next;
  }

  printf("Sum of even Fibonacci numbers <= 4,000,000: %d\n", sum);

  return 0;
}

File: ts/problem-2.ts
const MAX_NUMBER: number = 4000000;

function* fibonacciGenerator() : Iterator<number> {
  let prev: number = 0, cur: number = 1;
  let next: number;

  while (true) {
    next = prev + cur;
    if (next >= cur) {
      yield cur;
    } else {
      return cur;
    }
    
    prev = cur;
    cur = next;
    next = null;
  }
}

const sequence: Iterator<number> = fibonacciGenerator();
let sum_of_evens: number = 0;
let current_fib: IteratorResult<number>;

do {
  current_fib = sequence.next();
  if (current_fib.value % 2 === 0) {
    sum_of_evens += current_fib.value;
  }
} while (!current_fib.done && current_fib.value < MAX_NUMBER)

console.log('Sum of even Fibonacci numbers up to 4000000:', sum_of_evens);
